src/network/network_crypto.cpp

```diff
@@ -0,0 +1,112 @@
+/**
+ * Call \c crypto_wipe for all the data in the given span.
+ * @param span The span to cryptographically wipe.
+ */
+static void crypto_wipe(std::span<uint8_t> span)
+{
+	crypto_wipe(span.data(), span.size());
+}
+
+static ssize_t TransferOutSpan(std::span<uint8_t> destination, const char *source, size_t amount)
+{
+	size_t to_copy = std::min(amount, destination.size());
+	std::copy(source, source + to_copy, destination.data());
+	return to_copy;
+}
+
+template <size_t size>
+class SharedKeys {
+	std::array<uint8_t, 2 * size> keys;
+public:
+	~SharedKeys() { crypto_wipe(keys); }
+	constexpr std::span<const uint8_t> ClientToServer() const { return std::span(this->keys.data(), size); }
+	constexpr std::span<const uint8_t> ServerToClient() const { return std::span(this->keys.data() + size, size); }
+	constexpr std::span<uint8_t> Both() { return this->keys; }
+};
+
+X25519Key::~X25519Key()
+{
+	crypto_wipe(*this);
+}
+
+static bool X25519KeyExchange(SharedKeys<32> &shared_keys, const X25519Key &my_private_key, const X25519Key &their_public_key)
+{
+	X25519Key shared_secret;
+	crypto_x25519(shared_secret.data(), my_private_key.data(), their_public_key.data());
+	if (std::all_of(shared_secret.begin(), shared_secret.end(), [](auto v) { return v == 0; })) {
+		/* Secret is all 0 because public key is all 0, just reject it */
+		return false;
+	}
+
+	crypto_blake2b_ctx ctx;
+	crypto_blake2b_init(&ctx, shared_keys.Both().size());
+	crypto_blake2b_update(&ctx, shared_secret.data(), shared_secret.size());
+	crypto_blake2b_update(&ctx, my_public_key.data(), my_public_key.size());
+	crypto_blake2b_update(&ctx, their_public_key.data(), their_public_key.size());
+	crypto_blake2b_final(&ctx, shared_keys.Both().data());
+	return true;
+}
+
+
+X25519AuthorizedKeyHandler::X25519AuthorizedKeyHandler()
+{
+	RandomBytesWithFallback(this->my_private_key);
+	crypto_x25519_public_key(this->my_public_key.data(), this->my_private_key.data());
+
+	RandomBytesWithFallback(this->nonce);
+}
+
+X25519AuthorizedKeyHandler::~X25519AuthorizedKeyHandler()
+{
+	crypto_wipe(this->nonce);
+}
+
+/* virtual */ void X25519AuthorizedKeyHandler::SendRequest(Packet &p)
+{
+	p.Send_bytes(this->my_public_key);
+	p.Send_bytes(this->nonce);
+}
+
+/* virtual */ bool X25519AuthorizedKeyHandler::ReceiveResponse(Packet &p)
+{
+	X25519Key their_public_key{};
+	X25519Key message{};
+	std::array<uint8_t, 16> mac; // message authentication code
+
+	p.TransferOut(&TransferOutSpan, their_public_key);
+	p.TransferOut(&TransferOutSpan, message);
+	p.TransferOut(&TransferOutSpan, mac);
+
+	SharedKeys<32> shared_keys;
+	bool valid_exchange = X25519KeyExchange(shared_keys, this->my_private_key, their_public_key);
+	bool valid_unlock = crypto_aead_unlock(message.data(), mac.data(), shared_keys.ClientToServer().data(),
+			nonce.data(), their_public_key.data(), their_public_key.size(), message.data(), message.size()) == 0;
+	bool valid_message = crypto_verify32(message.data(), this->my_public_key.data());
```

Comment:
The message is already valid; `crypto_aead_unlock` promises that. So this is kinda redundant. There isn't really a payload to put in here that is needed.

Of course we can't send nothing, so something has to go in. And for that, I can see something like this. But it really isn't needed from a security point of view. Just making sure you know :D

Reply:
The message is valid, but is it what we expect the message to be?
At the client, I just inject some random bytes to encrypt and leave this check out. However, what are they going to put in there. If it works with all zero, then why not? Is that a situation we would want? I don't know.

JGRPP has an increasing counter which may not go backward in the payload, just to check it's not doing something fishy and the content is not always the same. Maybe to prevent replay attacks or something?

Also bear in mind that the primary reason I pushed was that I was at a road block for not being able to find the right "encrypt" function in the Python library. So I pushed what I had to ask around, but apparently doing something different for a while made me come up with a (nasty) solution in Python land (using two crypto libraries).

Reply:
Replay attacks was the main motivation for that, yes.

Reply:
We could make the step to encryption, as that would kinda make sense at this point. Then the payload can be the next packet the client sends to the server.

Just a random thought that came to mind.

Reply:
Given the server currently is always generating a new keypair, it's not feasible to do a replay attack. So I changed it to contain, for this authentication just some random data.